contentpal/src/main/java/org/dmfs/android/contentpal/predicates/arguments/BackReferenceArgument.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.android.contentpal.predicates.arguments;
+
+import android.support.annotation.NonNull;
+
+import org.dmfs.android.contentpal.Predicate;
+import org.dmfs.optional.Optional;
+import org.dmfs.optional.Present;
+
+
+/**
+ * @author Marten Gajda
+ */
+public final class BackReferenceArgument implements Predicate.Argument
+{
+    private final int mBackReference;
+
+
+    public BackReferenceArgument(int backReference)
+    {
+        mBackReference = backReference;
+    }
+
+
+    @NonNull
+    @Override
+    public String value()
+    {
+        return "-1";
```

Comment:
Doesn't this "-1" cause problem? Shouldn't this `Argument` only result in calling `builder.withSelectionBackReference()` and not in `builder.withSelection()`, but `BaseTable` will add this "-1" as well there.

Reply:
No, it's actually required. Here is how this works:

Assume the following (simplified) selection String: `completed = ? AND related_id = ?`. If you already know the related row id you would just pass the arguments array like `["1", "15"]`.

If you don't know the id (because the related row is added earlier in the same transaction) you still have to pass an array of two arguments. However, you can set a "back reference" which means, when this operation is to be executed, the argument at the given index is replaced with the id returned by the resulting id of the operation at the given index.

So if the batch contains 10 operations, the third one inserts a row which we refer to in the selection of the 8th (the one with the selection above), we would have to set the following back reference: 
```java
operation.withSelectionBackReference(
  1 /* 0-based index of the argument to replace */,
  2 /* 0-based of the operation in the batch, which created the row we refer to */
);
```

If the new row from the 3rd operation got the `id` `123`, the `ContentProviderOperation` would replace the `"15"` in the selection arguments array (at index 1) with `"123"` before this operation is executed.

So if everything goes well, the `-1` is replaced with the actual row id.

Reply:
I see, thanks for the explanation!